plugins/module_utils/ec2.py

```diff
@@ -315,10 +1204,419 @@ def normalize_ec2_vpc_dhcp_config(option_config):
+        if isinstance(instance_ids, string_types):
+            filters["instance-id"] = [instance_ids]
+        elif isinstance(instance_ids, list) and len(instance_ids) > 0:
+            filters["instance-id"] = instance_ids
+        else:
+            vpc_subnet_id = self.params.get("vpc_subnet_id")
+            filters["subnet-id"] = [vpc_subnet_id]
+            if not vpc_subnet_id:
+                network = self.params.get("network")
+                if network:
+                    # grab AZ from one of the ENIs
+                    interfaces = network.get("interfaces")
+                    if interfaces:
+                        filters["network-interface.network-interface-id"] = []
+                        for i in interfaces:
+                            if isinstance(i, dict):
+                                i = i["id"]
+                            filters["network-interface.network-interface-id"].append(i)
+                else:
+                    sub = (
+                        self.get_default_subnet(
+                            self.get_default_vpc() or {},
+                            availability_zone=self.params.get("availability_zone"),
+                        )
+                        or {}
+                    )
+                    filters["subnet-id"] = sub.get("SubnetId")
+
+            name = self.params.get("name")
+            tags = self.params.get("tags")
+            if name:
+                filters["tag:Name"] = [name]
+            elif tags:
+                name_tag = tags.get("Name")
+                if name_tag:
+                    filters["tag:Name"] = [name_tag]
+
+            image_id = self.params.get("image_id")
+            image = self.params.get("image")
+            if image_id:
+                filters["image-id"] = [image_id]
+            elif image and image.get("id"):
+                filters["image-id"] = [image.get("id")]
+        return filters
+
+    def find_instances(
+        self, ids: Optional[List[str]] = None, filters: Optional[Dict[str, Any]] = None
+    ) -> List[Dict[str, Any]]:
+        sanitized_filters = dict()
+        if ids:
+            params = {"InstanceIds": ids}
+        elif filters is None:
+            self.fail_json(msg="No filters provided when they were required")
+        else:
+            for key in list((filters or {}).keys()):
+                if not key.startswith("tag:"):
+                    sanitized_filters[key.replace("_", "-")] = (filters or {}).get(key)
+                else:
+                    sanitized_filters[key] = (filters or {}).get(key)
+            params = {"Filters": ansible_dict_to_boto3_filter_list(sanitized_filters)}
+
+        try:
+            results = []
+            for reservation in describe_instances(self.ec2, **params):
+                results += reservation.get("Instances", [])
+        except AnsibleEC2Error as e:
+            self.fail_json_aws(e, msg="Could not describe instances")
```

Comment:
Why catch our custom exceptions here? let them fall all the way back to the `main()` block and catch them there.  There's even `module.fail_json_aws_error()` which should handle things a little more cleanly.